rfc/2020/arm-support/aliariff.md

```diff
@@ -0,0 +1,267 @@
+     DOCKER_BUILD_ARCH=arm make build # building arm image
+     ```
+  - Push image & push multi arch manifest
+    ```yml
+    make push-all
+    ```
+
+- `linkerd2-proxy`
+  - Add new step in `.github/workflows/release.yml`
+  - Install requirement to perform cross compilation in Rust
+    ```yml
+    sudo apt-get install -y gcc-arm-linux-gnueabihf gcc-aarch64-linux-gnu
+    ```
+  - Set the appropriate environment
+    ```yml
+    env:
+      CARGO_TARGET_ARCH: armv7-unknown-linux-gnueabihf # for arm
+      CARGO_RELEASE: "1"
+    run: make package
+
+    or
+
+    env:
+      CARGO_TARGET_ARCH: aarch64-unknown-linux-gnu # for arm64
+      CARGO_RELEASE: "1"
+    run: make package
+    ```
+  - Change the Makefile to check for `CARGO_TARGET_ARCH` to perform cross compile
+    ```makefile
+    .
+    .
+    CARGO_BUILD = $(CARGO) build --frozen $(RELEASE) --target=$(CARGO_TARGET_ARCH)
+    .
+    .
+    ```
+  - Publish the artifact together with other architecture
+
+- `linkerd2`
+
+  There are 7 images to build: `controller`, `cli`, `cni-plugin`, `web`, `proxy`, `debug`, `grafana`
+
+  - Export `DOCKER_BUILD_ARCH` in `bin/_docker.sh` & append the architecture in the repo name
+    ```sh
+    export DOCKER_BUILD_ARCH="${DOCKER_BUILD_ARCH:-}"
+    .
+    .
+    .
+    docker_repo() {
+      .
+      .
+      .
+      if [ -n "${DOCKER_BUILD_ARCH:-}" ]; then
+        name="$name-$DOCKER_BUILD_ARCH"
+      fi
+      .
+      .
+      .
+    }
+    ```
+  - Pass the architecture argument to docker build in every `docker-build-*` file
+    ```sh
+    docker_build .... --build-arg "GOARCH=${DOCKER_BUILD_ARCH:-amd64}"
+    ```
+  - Use the `GOARCH` argument to perform cross compile in all `Dockerfile`
+    ```Dockerfile
+    ARG GOARCH
+    RUN CGO_ENABLED=0 GOOS=linux GOARCH=$GOARCH go build ...
+    ```
+  - Change the runtime image in `Dockerfile`
+    ```Dockerfile
+    FROM --platform=$GOARCH ...
+    ```
+  - Adjust `bin/fetch-proxy` file to retrieve the correct architecture version
+  - Add `bin/docker-manifest` file to create and push multi-arch manifest
+    ```sh
+    for img in cli-bin cni-plugin controller debug grafana proxy web ; do
+      docker manifest create gcr.io/linkerd-io/$img:$tag \
+        gcr.io/linkerd-io/$img-amd64:$tag \
+        gcr.io/linkerd-io/$img-arm64:$tag \
+        gcr.io/linkerd-io/$img-arm:$tag
+    done
+    ```
+  - Update `.github/workflows/release.yml`
+    ```yml
+    bin/docker-build
+    DOCKER_BUILD_ARCH=arm64 bin/docker-build
+    DOCKER_BUILD_ARCH=arm bin/docker-build
+    .
+    .
+    .
+    bin/docker-push $TAG
+    DOCKER_BUILD_ARCH=arm64 bin/docker-push $TAG
+    DOCKER_BUILD_ARCH=arm bin/docker-push $TAG
+
+    bin/docker-manifest $TAG
+    ```
+
+With all of step are explain, there is some drawback that we can see here as the we need to create manifest for multi-arch our self and every image architecture will have their own repositories.
```

Comment:
Last I tried, the `docker manifest` command is still experimental. I don't think there is a way to enable it on GitHub Actions.

> every image architecture will have their own repositories.

Can we assign different tags to the images, but store them in the same repository?

Reply:
I just try. It is possible to run experimental features on the Github Action runner.
![image](https://user-images.githubusercontent.com/6806035/77967176-e4812400-72e4-11ea-8738-f7b390299a55.png)
Ref: https://github.com/actions/virtual-environments/issues/368

Reply:
Different tags and the same repository are also possible. But that repo will look messy. But yeah totally possible. 
I create this by following how grafana team created it. Ref: https://hub.docker.com/u/grafana